packages/zksync-dai-bridge/contracts/l2/L2DAITokenBridge.sol

```diff
@@ -0,0 +1,133 @@
+}
+
+// Mint tokens on L2 after locking funds on L1.
+// Burn tokens on L1 and send a message to unlock tokens on L1 to L1 counterpart
+// Note: when bridge is closed it will still process in progress messages
+
+contract L2DAITokenBridge is IL2Bridge {
+    // --- Auth ---
+    mapping(address => uint256) public wards;
+
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "L2DAITokenBridge/not-authorized");
+        _;
+    }
+
+    address public immutable l2Token;
+    address public immutable l1Token;
+    address public immutable l1Bridge;
+    uint256 public isOpen = 1;
+
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Closed();
+
+    constructor(
+        address _l2Token,
+        address _l1Token,
+        address _l1Bridge
+    ) {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        l2Token = _l2Token;
+        l1Token = _l1Token;
+        l1Bridge = _l1Bridge;
+    }
+
+    function l2TokenAddress(address _l1Token) external view returns (address) {
+        require(_l1Token == l1Token, "L2DAITokenBridge/token-not-dai");
+        return l2Token;
+    }
+
+    function l1TokenAddress(address _l2Token) external view returns (address) {
+        require(_l2Token == l2Token, "L2DAITokenBridge/token-not-dai");
+        return l1Token;
+    }
+
+    function close() external auth {
+        isOpen = 0;
+
+        emit Closed();
+    }
+
+    function withdraw(
+        address _l1Receiver,
+        address _l2Token,
+        uint256 _amount
+    ) external {
+        require(_l2Token == l2Token, "L2DAITokenBridge/token-not-dai");
+
+        _initiateWithdrawal(_l1Receiver, _amount);
+
+        emit WithdrawalInitiated(msg.sender, _l1Receiver, _l2Token, _amount);
+    }
+
+    // When a withdrawal is initiated, we burn the withdrawer's funds to prevent subsequent L2 usage.
+    function _initiateWithdrawal(address _to, uint256 _amount) internal {
+        // do not allow initiating new xchain messages if bridge is closed
+        require(isOpen == 1, "L2DAITokenBridge/closed");
+
+        Mintable(l2Token).burn(msg.sender, _amount);
+
+        bytes memory message = abi.encodePacked(
+            IL1Bridge.finalizeWithdrawal.selector,
+            _to,
+            _amount
+        );
+
+        L1_MESSENGER_CONTRACT.sendToL1(message);
+    }
+
+    // When a deposit is finalized, we credit the account on L2 with the same amount of tokens.
+    function finalizeDeposit(
+        address _from,
+        address _to,
+        address _l1Token,
+        uint256 _amount,
+        bytes calldata /* _data */
+    ) external {
+        require(msg.sender == l1Bridge, "L2DAITokenBridge/sender-not-l1-bridge"); // only L1 bridge can call
```

Comment:
Optimism/Arbitrum uses address aliasing to avoid attack vector where the deployer of the L1 counterpart can spoof xchain messages (read more here: https://github.com/OffchainLabs/arbitrum/blob/master/docs/L1_L2_Messages.md#address-aliasing). 

Two solutions:
* enforce somehow that this is an xchain message (i am not sure if this is possible - depends on zksync implementation details). 
* burn deployer key's nonce so it's impossible to spoof such message